### PR TITLE
fix: two hook gaps — heredoc false positive and gh api contents bypass

### LINKED ISSUE #270: Two related gaps in the git-prism-redirect.sh hook
### Bug 1: Hook fires on gh pr diff appearing inside a heredoc body (false positive)

When a Bash tool call contains a heredoc whose body text mentions gh pr diff (e.g. as an example command in a GitHub issue description), the hook's tokenizer processes the heredoc body line-by-line and matches the literal text as a candidate command. The hook then fires a hard block even though no gh pr diff command is actually being executed.

*Repro:*

Create an issue whose body contains the text gh pr diff 123 --repo owner/repo inside a heredoc:

gh issue create --repo owner/repo \
  --title "Some title" \
  --body "$(cat <<'EOF'
Running:

 gh pr diff 123 --repo owner/repo

produces X behavior.
EOF
)"
The hook fires and blocks the entire gh issue create call, even though the gh pr diff text is inside a quoted heredoc string, not an executable command.

*Impact:* Any time an agent tries to create a GitHub issue or PR description that discusses or documents this hook (or gh pr diff in any example context), the hook blocks the action.

*Suggested fix:* Skip matching inside heredoc bodies. Lines between a <<'EOF' / <<EOF marker and the closing EOF delimiter should be excluded from candidate command extraction in _matches_gh_pr_diff (and the broader tokenize_command / _tokenize_raw pipeline if applicable).

---

### Bug 2: gh api repos/.../contents/...?ref=<sha> bypasses the hook (false negative)

The hook correctly intercepts gh pr diff and redirects to get_change_manifest / get_file_snapshots. However, fetching individual file snapshots from a PR ref via the GitHub REST contents endpoint is not intercepted:

gh api "repos/<owner>/<repo>/contents/<path>?ref=<head-sha>" --jq '.content' | base64 -d
This returns the full raw file content at a specific PR commit with no structured annotations and no hook feedback.

*Impact:* An agent (or user) can read any PR file outside of git-prism by fetching it at the PR's headRefOid. The result is unstructured text with no line-number context, losing the structured diff data that git-prism provides.

*Suggested fix:* Extend the hook to intercept gh api calls whose URL matches repos/<owner>/<repo>/contents/<path> when a ref query parameter is present (indicating a non-default-branch lookup tied to a PR ref). At that point, redirect to get_file_snapshots(repo, base_ref, head_ref, paths=[...]) as the structured alternative.

At minimum, update the hook's block message to note that gh api /contents/...?ref=<sha> is also a path to avoid, so agents know both escape routes are out of scope.

### PR DESCRIPTION
**Pull request by mikelane**
## Summary

Fixes two bugs in the bash redirect hook identified in #270:

- **Bug 1 (False positive):** `gh pr diff` inside heredoc bodies was incorrectly blocked when `<<` was inside a multi-line double-quoted string. Fixed by adding a regex pre-pass that strips heredoc bodies before tokenization.
- **Bug 2 (False negative):** `gh api repos/<owner>/<repo>/contents/<path>?ref=<sha>` bypassed the hook entirely. Fixed by detecting the contents-API-with-ref pattern and redirecting to `get_file_snapshots`.

## Commits
```
e5c3b4b fix: handle multiple heredocs on same line and malformed << without tag
c900c88 fix: allow digit-starting heredoc tags and match contents/ directly followed by ?ref=
018cfb0 fix: apply heredoc pre-pass to gh api contents interception
eae980e docs: document heredoc pre-pass and gh api contents interception
afccdd6 style: remove unused import io from bash_redirect_hook.py
d3de8fe fix: strip heredoc bodies from raw text and intercept gh api contents ref= calls
bb1a3ba test: add @ISSUE-270 BDD scenarios (RED)
```

## Test plan
- [x] 85 unit tests pass
- [x] 3 `@ISSUE-270` BDD scenarios pass
- [x] All 52 redirect_hook BDD scenarios pass (278 steps)
- [x] Adversarial QA (3 passes): 7 bugs found, all fixed
- [x] Church purification: python-purist, naming-purist, dead-code-purist, observability-purist

Closes #270

Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: hooks/bash_redirect_hook.py [hunk 1]
```diff
@@ -105,6 +113,60 @@ def _heredoc_tag(token: str) -> tuple[str, bool] | None:
         return None
     return tag, is_dash
```
Comment (mikelane) [inline hooks/bash_redirect_hook.py:115]: **HIGH**: `ruff format --check` fails on both changed Python files. These are formatting convention violations. Phase 1 quality tools must pass cleanly before merge.

```bash
$ python3 -m ruff format --check hooks/bash_redirect_hook.py hooks/test_bash_redirect_hook.py
Would reformat: hooks/bash_redirect_hook.py
Would reformat: hooks/test_bash_redirect_hook.py
2 files would be reformatted
```

**bash_redirect_hook.py** (two changes):
1. Missing blank line before module-level `_HEREDOC_START_PATTERN` constant
2. `except Exception:` clause formatting style

**test_bash_redirect_hook.py** (multiple changes):
- Multi-line heredoc body strings collapsed to single lines
- Function call argument formatting (single-line vs multi-line)
- Blank line adjustments between methods

Fix by running `python3 -m ruff format hooks/bash_redirect_hook.py hooks/test_bash_redirect_hook.py`.

File: hooks/test_bash_redirect_hook.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 from bash_redirect_hook import (
```
Comment (mikelane) [inline hooks/test_bash_redirect_hook.py:22]: **MEDIUM**: Tests directly import and exercise `_`-prefixed private module functions: `_advice_with_echo`, `_matches_gh_api_contents`, `_classify_git_command`, `_drop_heredoc_bodies`, `_has_pickaxe_flag`, `_has_ref_range`, `_is_functionally_empty`, `_matches_gh_pr_diff`.

The TDD rules require: "Test behavior via public methods only. Do not test private functions, internal state, or implementation details."

These are implementation-detail functions -- the BDD scenarios already cover the behavioral contract through the subprocess interface. Consider one of:
1. Remove the `_` prefix if these are intended as public API
2. Extract into a separate module (e.g., `tokenizer.py`) with a clean public API
3. Restructure unit tests to exercise edge cases through `decide_redirect()` / `tokenize_command()` only

File: hooks/bash_redirect_hook.py [hunk 3]
```diff
@@ -147,9 +212,11 @@ def _drop_heredoc_bodies(tokens: list[str]) -> list[str]:
                 _heredoc_tag(tokens[index + 1]) if index + 1 < len(tokens) else None
             )
             if tag_info is None:
-                # ``<<`` at end of input — drop the operator and stop.
-                break
-            tag, _is_dash = tag_info
+                # ``<<`` followed by an invalid/empty tag — skip the
+                # malformed operator and continue.
+                index += 1
```
Comment (mikelane) [inline hooks/bash_redirect_hook.py:218]: **MEDIUM**: `_drop_heredoc_bodies` (66 lines) and `_strip_heredocs_from_raw` (50 lines) implement state machines with deeply nested while loops. Consider extracting the body consumption logic into named helpers (e.g., `_consume_heredoc_body`) to clarify intent and keep each function under 20-30 lines.

The current code is correct with careful study, but nesting depth reaches 3 levels in places. Extracting named loops clarifies: "skip body lines until closing tag" vs "advance past body content tokens."

Comment (mikelane) [inline hooks/bash_redirect_hook.py:218]: **LOW**: When a malformed heredoc tag follows `<<` (e.g., `<< ''`), `index += 1; continue` skips the `<<` token but the invalid tag token (e.g., `''`) remains in the stream and gets appended via the normal-walk path.

This is harmless since an invalid tag won't match any redirect pattern, but to be thorough, advance past both the operator AND the invalid token: `index += 2; continue`.

Verification: the test at line 391 checks that `git` is preserved and `<<` is removed, but does not assert that the invalid token `''` is consumed.

**Review by mikelane — COMMENTED** [review]
## PR #272 Review: fix two hook gaps — heredoc false positive and gh api contents bypass

### Summary

This PR fixes two real bugs in the bash redirect hook from issue #270. The fixes are well-scoped, the test coverage is thorough (85 unit tests, 3 dedicated BDD scenarios, all 52 existing redirect_hook BDD scenarios still pass), and the adversarial QA uncovered 7 bugs that were all fixed before this review.

### Blocker

**HIGH — Formatter check failure**: `ruff format --check` fails on both changed Python files. Phase 1 quality tools must pass cleanly before merge. Fix by running:
```bash
python3 -m ruff format hooks/bash_redirect_hook.py hooks/test_bash_redirect_hook.py
```

### Recommendations

**MEDIUM — Private function testing**: 8 of 11 tested functions are `_`-prefixed module-level privates. Consider removing the `_` prefix or extracting into a dedicated tokenizer module.

**MEDIUM — State machine readability**: `_drop_heredoc_bodies` (66 lines) and `_strip_heredocs_from_raw` (50 lines) exceed the 20-line guideline. Extracting named helpers for the body-consumption loops would clarify intent.

**LOW — Incomplete invalid-tag skip**: In `_drop_heredoc_bodies`, malformed heredoc tags (line 214-218) skip the `<<` token but leak the invalid tag into the output stream. Harmless in practice but should advance past both tokens.

### What's Excellent

- **Commit structure**: 7 commits follow logical progression (RED BDD → core fix → style cleanup → docs → two reactive fixes from QA). Each commit has a clear `why` in the body.
- **Adversarial QA rigor**: 3 passes found 7 bugs. The reactive fixes (digit-starting tags, multiple heredocs on same line, heredoc pre-pass for gh api contents) demonstrate the process working as designed.
- **`_strip_heredocs_from_raw` + `_tokenize_raw` two-layer design**: The regex pre-pass catches heredocs inside quoted contexts that shlex can't tokenize, while the token-level `_drop_heredoc_bodies` handles remaining clean heredocs. This layers correctly — neither alone would solve the full problem.
- **CHANGELOG entries**: Clear, user-facing descriptions with issue references.
- **Regression safety**: Merge-base is origin/main HEAD — zero divergence, no sibling PR conflicts.

### Decision

The formatter failure (HIGH) must be resolved before merge. This is the only blocker. The remaining MEDIUM/LOW findings are recommendations, not gates.

Note: GitHub prohibits REQUEST_CHANGES on self-authored PRs. This COMMENT review is functionally equivalent.
